webui/react/src/hooks/useSettings.ts

```diff
@@ -1,359 +1,299 @@
+  userId: number;
 }
 
-export type UpdateSettings<T> = (newSettings: Partial<T>, push?: boolean) => void;
+export type UpdateSettings = (updates: Settings, shouldPush?: boolean) => Promise<void>;
+export type ResetSettings = (settings?: string[]) => void;
 
-export interface SettingsHook<T> {
+export type UseSettingsReturn<T> = {
   activeSettings: (keys?: string[]) => string[];
-  resetSettings: (keys?: string[]) => void;
-  settings: T;
-  updateSettings: UpdateSettings<T>;
-}
-
-export const validateBaseType = (type: BaseType, value: unknown): boolean => {
-  if (type === BaseType.Boolean && isBoolean(value)) return true;
-  if (type === BaseType.Float && isNumber(value)) return true;
-  if (type === BaseType.Integer && isNumber(value) && Math.ceil(value) === Math.floor(value))
-    return true;
-  if (type === BaseType.String && isString(value)) return true;
-  if (type === BaseType.Object && isObject(value)) return true;
-  return false;
+  isLoading: boolean;
+  resetSettings: ResetSettings;
+  settings: T | null;
+  updateSettings: UpdateSettings;
 };
 
-export const validateSetting = (config: SettingsConfigProp, value: unknown): boolean => {
-  if (value === undefined) return true;
-  if (config.type.isArray) {
-    if (!Array.isArray(value)) return false;
-    return value.every((val) => validateBaseType(config.type.baseType, val));
-  }
-  return validateBaseType(config.type.baseType, value);
-};
+const settingsToQuery = <T>(config: SettingsConfig<T>, settings: Settings) => {
+  const fullSettings = (Object.values(config.settings) as SettingsConfigProp<T>[]).reduce<Settings>(
+    (acc, setting) => {
+      // Save settings into query if there is value defined and is not the default value.
+      const value = settings[setting.storageKey];
+      const isDefault = isEqual(setting.defaultValue, value);
 
-export const getDefaultSettings = <T>(config: SettingsConfig, storage: Storage): T => {
-  return config.settings.reduce((acc, prop) => {
-    let defaultValue = prop.defaultValue;
-    if (prop.storageKey) {
-      defaultValue = storage.getWithDefault(prop.storageKey, defaultValue);
-    }
-    acc[prop.key] = defaultValue;
-    return acc;
-  }, {} as GenericSettings) as unknown as T;
+      acc[setting.storageKey] = !setting.skipUrlEncoding && !isDefault ? value : undefined;
+
+      return acc;
+    },
+    {},
+  );
+
+  return queryString.stringify(fullSettings);
 };
 
-export const queryParamToType = (type: BaseType, param: string | null): Primitive | undefined => {
-  if (param == null) return undefined;
+const queryParamToType = (type: BaseType, param: string | null): Primitive | undefined => {
+  if (param === null || param === undefined) return undefined;
   if (type === BaseType.Boolean) return param === 'true';
   if (type === BaseType.Float || type === BaseType.Integer) {
-    const value = type === BaseType.Float ? parseFloat(param) : parseInt(param);
+    const value = Number(param);
```

Comment:
this code can be replaced with https://github.com/sindresorhus/query-string#parsebooleans
i know this func exist before this PR, but I personally prefer to rely on lib funcs instead of making our own func if possible